File: 2026/06-June/24-Wednesday/solution.cpp
class Solution {
public:
    static const int MOD = 1e9 + 7;

    using Matrix = vector<vector<long long>>;

    Matrix multiply(const Matrix& A, const Matrix& B) {
        int n = A.size();
        Matrix C(n, vector<long long>(n, 0));

        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                if (A[i][k] == 0) continue;
                for (int j = 0; j < n; j++) {
                    C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % MOD;
                }
            }
        }
        return C;
    }

    Matrix power(Matrix base, long long exp) {
        int n = base.size();
        Matrix res(n, vector<long long>(n, 0));

        for (int i = 0; i < n; i++)
            res[i][i] = 1;

        while (exp) {
            if (exp & 1)
                res = multiply(res, base);

            base = multiply(base, base);
            exp >>= 1;
        }

        return res;
    }

    int zigZagArrays(int n, int l, int r) {
        int m = r - l + 1;

        Matrix U(m, vector<long long>(m, 0));
        Matrix D(m, vector<long long>(m, 0));

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < i; j++)
                U[i][j] = 1;

            for (int j = i + 1; j < m; j++)
                D[i][j] = 1;
        }

        auto UD = multiply(U, D);

        Matrix V = power(UD, (n - 1) / 2);

        if ((n - 1) % 2)
            V = multiply(V, U);

        long long ans = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < m; j++) {
                ans = (ans + V[i][j]) % MOD;
            }
        }

        return (ans * 2) % MOD;
    }
};

File: 2026/06-June/24-Wednesday/solution.py
class Solution:
    MOD = 10**9 + 7

    def multiply(self, A, B):
        n = len(A)
        C = [[0] * n for _ in range(n)]

        for i in range(n):
            for k in range(n):
                if A[i][k] == 0:
                    continue
                for j in range(n):
                    C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % self.MOD

        return C

    def power(self, base, exp):
        n = len(base)
        res = [[0] * n for _ in range(n)]

        for i in range(n):
            res[i][i] = 1

        while exp:
            if exp & 1:
                res = self.multiply(res, base)
            base = self.multiply(base, base)
            exp >>= 1

        return res

    def zigZagArrays(self, n, l, r):
        m = r - l + 1

        U = [[0] * m for _ in range(m)]
        D = [[0] * m for _ in range(m)]

        for i in range(m):
            for j in range(i):
                U[i][j] = 1
            for j in range(i + 1, m):
                D[i][j] = 1

        UD = self.multiply(U, D)

        V = self.power(UD, (n - 1) // 2)

        if (n - 1) % 2:
            V = self.multiply(V, U)

        ans = 0

        for row in V:
            ans = (ans + sum(row)) % self.MOD

        return (ans * 2) % self.MOD
